In [1]:
pip install tensorflow

     |████████████████████████████████| 320.4 MB 26 kB/s s eta 0:00:01|████▋                           | 45.9 MB 53.7 MB/s eta 0:00:065     |████████▊                       | 87.8 MB 59.9 MB/s eta 0:00:04     |██████████                      | 99.0 MB 59.9 MB/s eta 0:00:04MB/s eta 0:00:11██                    | 120.1 MB 20.9 MB/s eta 0:00:10     |████████████▍                   | 124.5 MB 20.9 MB/s eta 0:00:10��▊               | 167.9 MB 41.5 MB/s eta 0:00:04     |███████████████████▎            | 193.3 MB 21.9 MB/s eta 0:00:06     |████████████████████▋           | 206.9 MB 19.8 MB/s eta 0:00:06██████████████▊           | 208.0 MB 19.8 MB/s eta 0:00:06     |█████████████████████▎          | 213.3 MB 19.8 MB/s eta 0:00:06��███████████████████████▉      | 258.8 MB 37.1 MB/s eta 0:00:02279.3 MB 60.8 MB/s eta 0:00:01██████████████████▋   | 286.6 MB 60.8 MB/s eta 0:00:01��█████████████████▌  | 295.4 MB 60.8 MB/s eta 0:00:01ta 0:00:01     |██████████████████████████████  | 299.6 MB 60.8 MB/

Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install keras

Note: you may need to restart the kernel to use updated packages.


Load the data in pandas data frame.

In [21]:
import pandas as pd

url1="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_554093d294594a7e9051e3da088590a3.csv"
df1=pd.read_csv(url1,header=None)
df1.head()

0      1      2      3     4       5      6      7       8      9    ...  \
0    0   87.0  975.0 -287.0  11.0  1001.0  163.0   95.0   975.0  152.0  ...   
1   33  124.0  978.0 -389.0  -7.0  1014.0  199.0  124.0   968.0  123.0  ...   
2   67  102.0  996.0 -440.0 -49.0  1024.0  193.0  127.0  1001.0  113.0  ...   
3  100   59.0  861.0 -384.0  -9.0  1023.0  202.0  110.0  1007.0  106.0  ...   
4  133  119.0  946.0 -426.0 -22.0  1026.0  188.0   98.0  1001.0   92.0  ...   

      240     241     242  243  244  245  246  247  248  249  
0  5789.0  2907.0  1447.0    0    0    0    0    0    0    0  
1  5789.0  2908.0  1443.0    0    0    0    0    0    0    0  
2  5789.0  2910.0  1440.0    0    0    0    0    0    0    0  
3  5789.0  2912.0  1440.0    0    0    0    0    0    0    0  
4  5791.0  2915.0  1442.0    0    0    0    0    0    0    0  

[5 rows x 250 columns]

In [22]:
df1.shape

(51116, 250)

Separate the predictor variables (X) and target variables (y).

In [23]:
X=df1.iloc[:,0:243]
X.head()

0      1      2      3     4       5      6      7       8      9    ...  \
0    0   87.0  975.0 -287.0  11.0  1001.0  163.0   95.0   975.0  152.0  ...   
1   33  124.0  978.0 -389.0  -7.0  1014.0  199.0  124.0   968.0  123.0  ...   
2   67  102.0  996.0 -440.0 -49.0  1024.0  193.0  127.0  1001.0  113.0  ...   
3  100   59.0  861.0 -384.0  -9.0  1023.0  202.0  110.0  1007.0  106.0  ...   
4  133  119.0  946.0 -426.0 -22.0  1026.0  188.0   98.0  1001.0   92.0  ...   

      233     234     235     236     237     238     239     240     241  \
0  1265.0  5858.0  2707.0  1134.0  6294.0  2663.0  1533.0  5789.0  2907.0   
1  1267.0  5905.0  2708.0  1155.0  6294.0  2664.0  1532.0  5789.0  2908.0   
2  1268.0  5948.0  2710.0  1175.0  6295.0  2664.0  1531.0  5789.0  2910.0   
3  1266.0  5986.0  2711.0  1192.0  6296.0  2664.0  1530.0  5789.0  2912.0   
4  1263.0  6018.0  2712.0  1205.0  6299.0  2664.0  1529.0  5791.0  2915.0   

      242  
0  1447.0  
1  1443.0  
2  1440.0  
3  1440.0  
4  1442.0  

[5 rows x 243 columns]

In [24]:
y=df1.iloc[:,243:250]
y.head()

243  244  245  246  247  248  249
0    0    0    0    0    0    0    0
1    0    0    0    0    0    0    0
2    0    0    0    0    0    0    0
3    0    0    0    0    0    0    0
4    0    0    0    0    0    0    0

Check if the dataset has empty cells.

In [25]:
X.isna().sum()

0        0
1      911
2      911
3      911
4      466
      ... 
238    409
239    409
240    409
241    409
242    409
Length: 243, dtype: int64

It looks like there are a lot of empty cells in the predictor variables.

In [26]:
y.isna().sum()

243    0
244    0
245    0
246    0
247    0
248    0
249    0
dtype: int64

There is no empty cells in the target variables.

Calculate the percentage of empty cells in each column in the predictor variables.

In [27]:
p_m =pd.DataFrame(X.isnull().sum() * 100 / len(X)).rename(columns={0:"Percentage"})
p_m=p_m.sort_values(by="Percentage",ascending=False)
p_m

Percentage
36    99.998044
35    99.998044
34    99.998044
219   99.998044
220   99.998044
..          ...
203    0.000000
202    0.000000
201    0.000000
200    0.000000
0      0.000000

[243 rows x 1 columns]

It appears that some columns have as much as 99.99% of empty cells. These columns will be discarded. We will discard columns with more than 20% of missing data.

In [28]:
p_m=p_m[p_m["Percentage"]<=20]
p_m

Percentage
20     4.067220
21     4.067220
19     4.067220
15     2.369121
13     2.369121
..          ...
203    0.000000
202    0.000000
201    0.000000
200    0.000000
0      0.000000

[237 rows x 1 columns]

After discarding columns with more than 20% of missing values, we have 237 columns in the predictor variables X.

Select those columns with at most 20% missing values.

In [29]:
X=X.iloc[:,p_m.index.tolist()]
X.head()

20    21     19     15     13      14     143   139    140    141  ...  \
0  973.0 -41.0  209.0   99.0  249.0  1008.0  411.0 -72.0  682.0 -778.0  ...   
1  956.0 -93.0  245.0   97.0  272.0   963.0  409.0 -72.0  684.0 -770.0  ...   
2  929.0 -69.0  277.0  116.0  213.0  1017.0  410.0 -72.0  684.0 -770.0  ...   
3  958.0 -76.0  245.0  120.0  234.0  1028.0  411.0 -69.0  673.0 -773.0  ...   
4  968.0 -20.0  231.0  189.0  191.0   991.0  411.0 -62.0  684.0 -787.0  ...   

   196  197  206  205  204  203  202  201  200  0    
0    0    0    0    0    0    0    0    0    0    0  
1    0    0    0    0    0    0    0    0    0   33  
2    0    0    0    0    0    0    0    0    0   67  
3    0    0    0    0    0    0    0    0    0  100  
4    0    0    0    0    0    0    0    0    0  133  

[5 rows x 237 columns]

Fill those empty cells with the median of the respective column.

In [30]:
X=X.fillna(X.median())

Verified that there is no more missing values in the predictor variables X.

In [31]:
null=pd.DataFrame(X.isna().any()).rename(columns={0:"Null"})
null[null["Null"]==True]

Empty DataFrame
Columns: [Null]
Index: []

All empty cells have been filled.

For this task, we will build a model to predict the locomotion: Stand, Walk, Sit or Lie. That is the first column in the target variables y.

In [32]:
y1=pd.DataFrame(y.iloc[:,0]).rename(columns={243:"locomotion"})
y1

locomotion
0               0
1               0
2               0
3               0
4               0
...           ...
51111           0
51112           0
51113           0
51114           0
51115           0

[51116 rows x 1 columns]

In [33]:
import numpy as np
np.unique(y1)

array([0, 1, 2, 4, 5])

Feature extraction.

In [34]:
from sklearn import preprocessing
scaler=preprocessing.StandardScaler()
X=scaler.fit_transform(X)
X=pd.DataFrame(X)
X.head()

0         1         2         3         4         5         6    \
0  0.290387  0.493052 -0.115890  0.205712 -0.433329  0.590983 -0.045576   
1  0.224739  0.265039  0.001353  0.200330 -0.356390  0.508880 -0.645468   
2  0.120474  0.370276  0.105569  0.251459 -0.553755  0.607403 -0.345522   
3  0.232462  0.339582  0.001353  0.262223 -0.483506  0.627473 -0.045576   
4  0.271079  0.585134 -0.044241  0.447903 -0.627348  0.559966 -0.045576   

        7         8         9    ...       227       228       229       230  \
0 -0.189098  0.018619 -0.408523  ... -0.056035 -0.024634 -0.020273 -0.019283   
1 -0.189098  0.026657 -0.384108  ... -0.056035 -0.024634 -0.020273 -0.019283   
2 -0.189098  0.026657 -0.384108  ... -0.056035 -0.024634 -0.020273 -0.019283   
3 -0.156555 -0.017552 -0.393263  ... -0.056035 -0.024634 -0.020273 -0.019283   
4 -0.080623  0.026657 -0.435990  ... -0.056035 -0.024634 -0.020273 -0.019283   

        231       232       233  234  235       236  
0 -0.158326 -0.021674 -0.023826  0.0  0.0 -1.732017  
1 -0.158326 -0.021674 -0.023826  0.0  0.0 -1.731950  
2 -0.158326 -0.021674 -0.023826  0.0  0.0 -1.731881  
3 -0.158326 -0.021674 -0.023826  0.0  0.0 -1.731814  
4 -0.158326 -0.021674 -0.023826  0.0  0.0 -1.731747  

[5 rows x 237 columns]

In [35]:
from sklearn.decomposition import PCA, IncrementalPCA

ipca = IncrementalPCA(n_components=30,whiten=True)
X=ipca.fit_transform(X)

In [36]:
X=pd.DataFrame(X)
X

0         1         2         3         4         5         6   \
0     -0.779637 -0.478474  0.635114  1.090687 -0.255061 -1.231526  0.308716   
1     -0.776115 -0.493976  0.653046  1.127359 -0.228167 -1.325151 -0.062817   
2     -0.766890 -0.416995  0.584813  1.148440 -0.254395 -1.278737 -0.426898   
3     -0.776255 -0.442681  0.643475  1.159379 -0.272868 -1.333967 -0.416260   
4     -0.770985 -0.442353  0.673203  1.066249 -0.271848 -1.307775 -0.385441   
...         ...       ...       ...       ...       ...       ...       ...   
51111 -0.282918 -0.334041  0.058231  0.151783  0.058806 -0.165540  0.031032   
51112 -0.282916 -0.334039  0.058229  0.151782  0.058805 -0.165539  0.031031   
51113 -0.282915 -0.334036  0.058227  0.151781  0.058804 -0.165538  0.031030   
51114 -0.282914 -0.334034  0.058225  0.151781  0.058803 -0.165537  0.031030   
51115 -0.282912 -0.334031  0.058223  0.151780  0.058803 -0.165537  0.031029   

             7         8         9   ...        20        21        22  \
0     -1.359427 -0.128528  0.105931  ...  1.007063  0.521752  0.060881   
1     -1.399275  0.053751 -0.021489  ...  1.067392  0.577343 -0.017322   
2     -1.380744  0.067258  0.047619  ...  0.983196  0.226752  0.091084   
3     -1.423296 -0.232984  0.035836  ...  1.063633 -0.078714  0.009909   
4     -1.337149  0.262240  0.053162  ...  0.877758 -0.170506  0.103188   
...         ...       ...       ...  ...       ...       ...       ...   
51111  0.230290  0.095166 -0.006590  ...  0.207980  0.004780  0.047068   
51112  0.230293  0.095166 -0.006590  ...  0.207979  0.004781  0.047070   
51113  0.230296  0.095165 -0.006590  ...  0.207978  0.004782  0.047073   
51114  0.230299  0.095165 -0.006590  ...  0.207977  0.004783  0.047075   
51115  0.230302  0.095165 -0.006590  ...  0.207976  0.004783  0.047077   

             23        24        25        26        27        28        29  
0      1.090612 -0.072613 -0.004863 -0.050337  0.936883  0.045683  0.010402  
1      1.366246 -0.411713 -0.069551  0.047967  0.556898 -0.134072  0.365830  
2      1.507945 -0.030696 -0.237545 -0.129532  0.374661 -0.247567  0.284645  
3      1.220695  0.086734  0.026009 -0.413850  0.109655  0.417410 -0.030357  
4      1.103431 -0.060818 -0.110982 -0.224408  0.126037  0.152301 -0.220722  
...         ...       ...       ...       ...       ...       ...       ...  
51111 -0.037535  0.238295 -0.214540  0.018184 -0.027917  0.061008 -0.411142  
51112 -0.037535  0.238296 -0.214541  0.018184 -0.027918  0.061006 -0.411144  
51113 -0.037535  0.238298 -0.214542  0.018184 -0.027918  0.061005 -0.411145  
51114 -0.037535  0.238299 -0.214543  0.018184 -0.027919  0.061004 -0.411146  
51115 -0.037535  0.238301 -0.214545  0.018184 -0.027919  0.061003 -0.411148  

[51116 rows x 30 columns]

Check for class imbalance.

In [37]:
locomotion=pd.concat([X,y1],axis=1)
(locomotion.groupby("locomotion").count().T)/len(locomotion)

locomotion         0         1         2         4         5
0           0.266238  0.437828  0.128003  0.145864  0.022067
1           0.266238  0.437828  0.128003  0.145864  0.022067
2           0.266238  0.437828  0.128003  0.145864  0.022067
3           0.266238  0.437828  0.128003  0.145864  0.022067
4           0.266238  0.437828  0.128003  0.145864  0.022067
5           0.266238  0.437828  0.128003  0.145864  0.022067
6           0.266238  0.437828  0.128003  0.145864  0.022067
7           0.266238  0.437828  0.128003  0.145864  0.022067
8           0.266238  0.437828  0.128003  0.145864  0.022067
9           0.266238  0.437828  0.128003  0.145864  0.022067
10          0.266238  0.437828  0.128003  0.145864  0.022067
11          0.266238  0.437828  0.128003  0.145864  0.022067
12          0.266238  0.437828  0.128003  0.145864  0.022067
13          0.266238  0.437828  0.128003  0.145864  0.022067
14          0.266238  0.437828  0.128003  0.145864  0.022067
15          0.266238  0.437828  0.128003  0.145864  0.022067
16          0.266238  0.437828  0.128003  0.145864  0.022067
17          0.266238  0.437828  0.128003  0.145864  0.022067
18          0.266238  0.437828  0.128003  0.145864  0.022067
19          0.266238  0.437828  0.128003  0.145864  0.022067
20          0.266238  0.437828  0.128003  0.145864  0.022067
21          0.266238  0.437828  0.128003  0.145864  0.022067
22          0.266238  0.437828  0.128003  0.145864  0.022067
23          0.266238  0.437828  0.128003  0.145864  0.022067
24          0.266238  0.437828  0.128003  0.145864  0.022067
25          0.266238  0.437828  0.128003  0.145864  0.022067
26          0.266238  0.437828  0.128003  0.145864  0.022067
27          0.266238  0.437828  0.128003  0.145864  0.022067
28          0.266238  0.437828  0.128003  0.145864  0.022067
29          0.266238  0.437828  0.128003  0.145864  0.022067

The classes that represent the motions are not evenly distributed. Almost 44% of the data belongs to Class 1, while only slightly more than 2% belongs to Class 5. We are going to oversample the minority classes (Classes 2,4,5): duplicate Classes 2 and 4 by 2 times and Class 5 by 10 times.

To ensure that test data consists of strictly data that the model has not seen during the training process, we will split the data between oversampling. In other words, we will only oversample the minority classes in the train data.

In [56]:
locomotion_train=locomotion.sample(frac=0.9)
locomotion_train.shape

(46004, 31)

In [57]:
print(51116*0.9)
print(51116*0.1)

46004.4
5111.6


In [58]:
L_0=locomotion_train[locomotion_train["locomotion"]==0]
L_1=locomotion_train[locomotion_train["locomotion"]==1]
L_2=locomotion_train[locomotion_train["locomotion"]==2]
L_4=locomotion_train[locomotion_train["locomotion"]==4]


L_5=locomotion_train[locomotion_train["locomotion"]==5]
L_5=pd.concat([L_5]*5,axis=0)

In [59]:
locomotion_train=pd.concat([L_0,L_1,L_2,L_4,L_5],axis=0)
locomotion_train.shape

(50056, 31)

In [60]:
(locomotion_train.groupby("locomotion").count().T)/len(locomotion_train)

locomotion         0        1         2        4         5
0           0.245325  0.40149  0.118268  0.13373  0.101187
1           0.245325  0.40149  0.118268  0.13373  0.101187
2           0.245325  0.40149  0.118268  0.13373  0.101187
3           0.245325  0.40149  0.118268  0.13373  0.101187
4           0.245325  0.40149  0.118268  0.13373  0.101187
5           0.245325  0.40149  0.118268  0.13373  0.101187
6           0.245325  0.40149  0.118268  0.13373  0.101187
7           0.245325  0.40149  0.118268  0.13373  0.101187
8           0.245325  0.40149  0.118268  0.13373  0.101187
9           0.245325  0.40149  0.118268  0.13373  0.101187
10          0.245325  0.40149  0.118268  0.13373  0.101187
11          0.245325  0.40149  0.118268  0.13373  0.101187
12          0.245325  0.40149  0.118268  0.13373  0.101187
13          0.245325  0.40149  0.118268  0.13373  0.101187
14          0.245325  0.40149  0.118268  0.13373  0.101187
15          0.245325  0.40149  0.118268  0.13373  0.101187
16          0.245325  0.40149  0.118268  0.13373  0.101187
17          0.245325  0.40149  0.118268  0.13373  0.101187
18          0.245325  0.40149  0.118268  0.13373  0.101187
19          0.245325  0.40149  0.118268  0.13373  0.101187
20          0.245325  0.40149  0.118268  0.13373  0.101187
21          0.245325  0.40149  0.118268  0.13373  0.101187
22          0.245325  0.40149  0.118268  0.13373  0.101187
23          0.245325  0.40149  0.118268  0.13373  0.101187
24          0.245325  0.40149  0.118268  0.13373  0.101187
25          0.245325  0.40149  0.118268  0.13373  0.101187
26          0.245325  0.40149  0.118268  0.13373  0.101187
27          0.245325  0.40149  0.118268  0.13373  0.101187
28          0.245325  0.40149  0.118268  0.13373  0.101187
29          0.245325  0.40149  0.118268  0.13373  0.101187

In [61]:
locomotion_test=locomotion.drop(locomotion.index[locomotion_train.index.tolist()])
locomotion_test.shape

(5112, 31)

In [62]:
(locomotion_test.groupby("locomotion").count().T)/len(locomotion_test)

locomotion         0         1        2         4         5
0           0.259977  0.446596  0.12187  0.149061  0.022496
1           0.259977  0.446596  0.12187  0.149061  0.022496
2           0.259977  0.446596  0.12187  0.149061  0.022496
3           0.259977  0.446596  0.12187  0.149061  0.022496
4           0.259977  0.446596  0.12187  0.149061  0.022496
5           0.259977  0.446596  0.12187  0.149061  0.022496
6           0.259977  0.446596  0.12187  0.149061  0.022496
7           0.259977  0.446596  0.12187  0.149061  0.022496
8           0.259977  0.446596  0.12187  0.149061  0.022496
9           0.259977  0.446596  0.12187  0.149061  0.022496
10          0.259977  0.446596  0.12187  0.149061  0.022496
11          0.259977  0.446596  0.12187  0.149061  0.022496
12          0.259977  0.446596  0.12187  0.149061  0.022496
13          0.259977  0.446596  0.12187  0.149061  0.022496
14          0.259977  0.446596  0.12187  0.149061  0.022496
15          0.259977  0.446596  0.12187  0.149061  0.022496
16          0.259977  0.446596  0.12187  0.149061  0.022496
17          0.259977  0.446596  0.12187  0.149061  0.022496
18          0.259977  0.446596  0.12187  0.149061  0.022496
19          0.259977  0.446596  0.12187  0.149061  0.022496
20          0.259977  0.446596  0.12187  0.149061  0.022496
21          0.259977  0.446596  0.12187  0.149061  0.022496
22          0.259977  0.446596  0.12187  0.149061  0.022496
23          0.259977  0.446596  0.12187  0.149061  0.022496
24          0.259977  0.446596  0.12187  0.149061  0.022496
25          0.259977  0.446596  0.12187  0.149061  0.022496
26          0.259977  0.446596  0.12187  0.149061  0.022496
27          0.259977  0.446596  0.12187  0.149061  0.022496
28          0.259977  0.446596  0.12187  0.149061  0.022496
29          0.259977  0.446596  0.12187  0.149061  0.022496

Because the dataset was split first and then only the train data was oversampled, although the proportions of the classes change in the train data, the proportions of classes in the test data remain the same.

In [63]:
X_train=locomotion_train.drop(["locomotion"],axis=1)
y_train=locomotion_train[["locomotion"]]

In [64]:
X_test=locomotion_test.drop(["locomotion"],axis=1)
y_test=locomotion_test[["locomotion"]]

In [65]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
y_train=pd.DataFrame(enc.fit_transform(y_train).toarray())
y_train

0    1    2    3    4
0      1.0  0.0  0.0  0.0  0.0
1      1.0  0.0  0.0  0.0  0.0
2      1.0  0.0  0.0  0.0  0.0
3      1.0  0.0  0.0  0.0  0.0
4      1.0  0.0  0.0  0.0  0.0
...    ...  ...  ...  ...  ...
50051  0.0  0.0  0.0  0.0  1.0
50052  0.0  0.0  0.0  0.0  1.0
50053  0.0  0.0  0.0  0.0  1.0
50054  0.0  0.0  0.0  0.0  1.0
50055  0.0  0.0  0.0  0.0  1.0

[50056 rows x 5 columns]

In [66]:
y_test=pd.DataFrame(enc.fit_transform(y_test).toarray())
y_test

0    1    2    3    4
0     1.0  0.0  0.0  0.0  0.0
1     1.0  0.0  0.0  0.0  0.0
2     1.0  0.0  0.0  0.0  0.0
3     1.0  0.0  0.0  0.0  0.0
4     1.0  0.0  0.0  0.0  0.0
...   ...  ...  ...  ...  ...
5107  1.0  0.0  0.0  0.0  0.0
5108  1.0  0.0  0.0  0.0  0.0
5109  1.0  0.0  0.0  0.0  0.0
5110  1.0  0.0  0.0  0.0  0.0
5111  1.0  0.0  0.0  0.0  0.0

[5112 rows x 5 columns]

In [67]:
# Load the libraries
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import Adam

In [68]:
length=X_train.shape[1]
num_classes=y_test.shape[1]

def classification_model():
    # create model
    model = Sequential()
    model.add(Dense(length, activation='relu', input_shape=(length,)))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(60, activation='relu'))
    model.add(Dense(60, activation= 'relu'))
    model.add(Dense(60, activation = 'relu'))
    model.add(Dense(num_classes, activation='softmax'))
    opt=Adam(lr=0.001)
    model.compile(optimizer=opt, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

In [70]:
model=classification_model()
model.fit(X_train,y_train,validation_data=(X_test, y_test),epochs=50, verbose=2)
model.evaluate(X_test,y_test,verbose=0)

Epoch 1/50
1565/1565 - 7s - loss: 0.3142 - accuracy: 0.8805 - val_loss: 0.2272 - val_accuracy: 0.9216
Epoch 2/50
1565/1565 - 6s - loss: 0.1667 - accuracy: 0.9384 - val_loss: 0.1613 - val_accuracy: 0.9419
Epoch 3/50
1565/1565 - 6s - loss: 0.1294 - accuracy: 0.9519 - val_loss: 0.1376 - val_accuracy: 0.9509
Epoch 4/50
1565/1565 - 6s - loss: 0.1083 - accuracy: 0.9607 - val_loss: 0.1413 - val_accuracy: 0.9499
Epoch 5/50
1565/1565 - 6s - loss: 0.0936 - accuracy: 0.9669 - val_loss: 0.1137 - val_accuracy: 0.9605
Epoch 6/50
1565/1565 - 6s - loss: 0.0814 - accuracy: 0.9711 - val_loss: 0.1244 - val_accuracy: 0.9587
Epoch 7/50
1565/1565 - 6s - loss: 0.0745 - accuracy: 0.9741 - val_loss: 0.1020 - val_accuracy: 0.9640
Epoch 8/50
1565/1565 - 6s - loss: 0.0670 - accuracy: 0.9766 - val_loss: 0.1052 - val_accuracy: 0.9630
Epoch 9/50
1565/1565 - 7s - loss: 0.0589 - accuracy: 0.9786 - val_loss: 0.1020 - val_accuracy: 0.9646
Epoch 10/50
1565/1565 - 8s - loss: 0.0560 - accuracy: 0.9799 - val_loss: 0.0872 - 

[0.07750649005174637, 0.9825899600982666]